llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp

```diff
@@ -2585,6 +2585,8 @@ SDValue DAGTypeLegalizer::PromoteIntOp_ExpOp(SDNode *N) {
                              : RTLIB::getLDEXP(N->getValueType(0));
 
   if (LC == RTLIB::UNKNOWN_LIBCALL || !TLI.getLibcallName(LC)) {
+    if (N->getValueType(0).isVector())
+      return DAG.UnrollVectorOp(N);
```

Comment:
This whole function is actually the wrong place to split the vector (see how there are no other UnrollVectorOps uses in DAGTypeLegalizer). The description also says your problem is when the libcall is used, so you'd want to change the other path? Does the libcall emission below need to directly handle the vector case?

Reply:
Sorry, I'm not entirely sure I understand what you mean. Could you please clarify? 
Are you referring to the approach used in my first commit https://github.com/llvm/llvm-project/pull/118242/commits/7f5a128d9c3f9468f7a4c6bd2c08dfa478f3bf5c ? If so, in that commit, you mentioned not to check for the `SIGN_EXTEND_INREG` node, but this node is generated in this `if `statement here.(`SExtPromotedInteger` will generate a `SIGN_EXTEND_INREG` node). 
```
  if (LC == RTLIB::UNKNOWN_LIBCALL || !TLI.getLibcallName(LC)) {
    SmallVector<SDValue, 3> NewOps(N->ops());
    NewOps[1 + OpOffset] = SExtPromotedInteger(N->getOperand(1 + OpOffset));
    return SDValue(DAG.UpdateNodeOperands(N, NewOps), 0);
  }
```

Other info:
1. RISCV64 and Loongarch64 will enter function `PromoteIntOp_ExpOp` and generate a `SIGN_EXTEND_INREG` node, so `ExponentHasSizeOfInt=false`(32 != 64) and report an error.
```
bool ExponentHasSizeOfInt =
        DAG.getLibInfo().getIntSize() ==
        Node->getOperand(1 + Offset).getValueType().getSizeInBits();
```
Additionally, AArch64 and X86 will get `ExponentHasSizeOfInt=true`, because in these backends, i32 is valid and was not promoted to `SIGN_EXTEND_INREG`(i64) previously. So, they generate right code sequences.

2. In the entire LLVM, only `fpowi` and `fldexp` will call the `PromoteIntOp_ExpOp` function.(Flow: `llvm::DAGTypeLegalizer::PromoteIntegerOperand` -> `llvm::DAGTypeLegalizer::PromoteIntOp_ExpOp`)
```
  case ISD::FPOWI:
  case ISD::STRICT_FPOWI:
  case ISD::FLDEXP:
  case ISD::STRICT_FLDEXP: Res = PromoteIntOp_ExpOp(N); break;
```

If not, could you provide more specific modification suggestions？

Reply:
> see how there are no other UnrollVectorOps uses in DAGTypeLegalizer

There are calls to `UnrollVectorOps` in several places in LegalizeVectorTypes.cpp

This comment where the libcall below is created seems relevant

```
  // We can't just promote the exponent type in FPOWI, since we want to lower
  // the node to a libcall and we if we promote to a type larger than
  // sizeof(int) the libcall might not be according to the targets ABI.
```

My suggestion to unroll here was so that we wouldn't promote past sizeof(int). If we wait until LegalizeDAG.cpp to unroll the operation, the damage to the integer type has already been done. For RISC-V its harmless because `signed int` is supposed to be passed sign extended to 64-bits according to the ABI.

Hypothetically, if powi took an `unsigned int` as an argument, then type legalization would use zero extend, but the RISC-V ABI wants `unsigned int` to be passed sign extended. So LegalizeDAG would need to insert a SIGN_EXTEND_INREG to fix. I guess it would need to use the getIntSize() and shouldSignExtendTypeInLibCall to know what it needs to do in that case.

If we don't unroll here I guess the best fix in LegalizeDAG would also be to use `getIntSize()` and `shouldSignExtendTypeInLibCall` and use computeNumSignBits to know if a SIGN_EXTEND_INREG needs to be inserted?

Reply:
It makes more sense to me to handle this when emitting the call, where ABI constraints would naturally be handled. This can be sign extended here, and the call emission can truncate or sext_inreg as required

Reply:
> It makes more sense to me to handle this when emitting the call, where ABI constraints would naturally be handled. This can be sign extended here, and the call emission can truncate or sext_inreg as required

The POWI code in LegalizeDAG calls `SelectionDAGLegalize::ExpandFPLibCall` which  will call `TargetLowering::makeLibCall` using the promoted type. `makeLibCall` calls `getTypeForEVT` which will return `i64` due to the promotion. That's what will be used by calling lowering, but that's the wrong type for it to do the right thing. We need to get an i32 Type* into call lowering, but we no longer have it. We'll need to call getIntSize() to get the size and pass it along somehow. That requires refactoring several interfaces or adding new ones. Not sure if call lowering would also expect the SDValue for the argument to have i32 type.

My unrolling proposal avoided that by scalarizing it and letting the newly created scalar powi calls get converted to libcalls while we're still in type legalization. That's how we currently handle the scalar powi case. Are you also suggesting we should move the scalar handling to LegalizeDAG as well?

Reply:
makeLibCall should really be looking at the signature of the underlying call, but RuntimeLibcalls currently does not record this information. TargetLibraryInfo does, which is separate for some reason. This keeps coming up as a problem, these really need to be merged in some way (cc @jhuber6).

Taking the type from the DAG node isn't strictly correct, it's just where we've ended up. This came up recently for the special case in ExpandFPLibCall to sign extend the integer argument for FLDEXP. 

Practically speaking, I don't think any targets will have a vector powi implementation (I at least don't see any in RuntimeLibcalls), so unrolling works out. I guess this could get a fixme and go ahead for now, but it's still a hack